rfcs/20220329-Extending-CUDA-backend-of-oneDNN-Primitives/README.md

```diff
@@ -0,0 +1,190 @@
+
+- This branch contains the support for primitives that are yet to be supported for CUDA backend. 
+
+- Compilation and functional validation is yet to be concluded.  
+
+## Proposal 
+
+The primitives are built using oneDNN and are the open-source DNN libraries of Intel. 
+
+ - This contribution would extend oneDNN's support for more primitives in CUDA backend. 
+
+ - Wherever direct mapping of primitives from oneDNN is not available in cuDNN, the implementation would proceed with custom kernels for emulating their functionality on the backend device. 
+
+ 
+
+## Supported Primitives,  Implementation Limitations and Proposed Solution 
+
+ 
+
+### Reduction 
+
+- Reduction primitive in CUDA backend can be implemented using cudnnReduceTensor primitive of cuDNN.
+- cudnnReduceTensor supports only five modes of reduction operations via enumerators which are as follows.  
+    - `CUDNN_REDUCE_TENSOR_MAX` 
+    - `CUDNN_REDUCE_TENSOR_MIN` 
+    - `CUDNN_REDUCE_TENSOR_MUL`
+    - `CUDNN_REDUCE_TENSOR_ADD` 
+    - `CUDNN_REDUCE_TENSOR_AVG` 
+
+- Supported Data Types:  f32, s8 
+
+ 
+
+#### Implementation Limitations 
+
+- Lack of support for the algorithms of  
+
+    - `reduction_norm_lp_max` 
+
+    - `reduction_norm_lp_sum` 
+
+    - `reduction_norm_lp_power_p_max` 
+
+    - `reduction_norm_lp_power_p_sum` 
+
+- No support for bf16 datatype. 
+
+ 
+
+#### Proposed Solutions 
+
+- To achieve the functionality of `reduction_norm_lp_power_p_max` and `reduction_norm_lp_power_p_sum` algorithms in cuDNN, `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` can be used in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 1.  
+
+- Similarly, the functionality of `reduction_norm_lp_max` and `reduction_norm_lp_sum` algortihms can be implemented using `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 2.  
+
+- As an alternative to the above workarounds, the functionality of the above algorithms can be implemented through a custom kernel in SYCL. 
+
+ 
+
+### Recurrent Neural Networks 
+
+- RNN primitive in CUDA backend can be implemented using `cudnnRNNForward` and `cudnnRNNBackward` primitives of cuDNN. 
+
+- cuDNN's primitives only support two algorithm of RNN. 
+
+    - `CUDNN_LSTM` 
+
+    - `CUDNN_GRU` 
+
+- Supported Data Types: f16, f32 
+
+ 
+
+#### Propagation 
+
+Supported propagation kinds: Forward Training/Inference and Backward 
+
+ 
+
+#### Implementation Limitations 
+
+- No support for following algorithms in cuDNN. 
+
+    - `augru` 
+
+    - `lbr_augru` 
+
+    - `lbr_gru` 
```

Comment:
Actually cuDNN GRU follows oneDNN LBR_GRU:
cuDNN doc: https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnRNNMode_t
```
i t   =   σ   ( W i x t   +   R i h t - 1   + b Wi   +   b Ru )
r t   =   σ   ( W r x t   +   R r h t - 1   + b Wr   +   b Rr )
h ʹ t   =   tanh   ( W h x t   +   r t   ◦   ( R h h t-1   + b Rh )   +   b Wh )
h t   =   ( 1 - i t )   ◦   h ʹ t   +   i t   ◦ h t-1
```
As you can see linear operation is applied before reset gate (r t) which is called LBR in oneDNN.

Reply:
Sure. We explored that in cuDNN,  there is a bias mode cudnnRNNBiasMode_t. By altering the values between CUDNN_RNN_SINGLE_INP_BIAS and CUDNN_RNN_DOUBLE_BIAS we can achieve GRU and LBR_GRU respectively. 

 https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnRNNBiasMode_t